Added gbt()  and converted exist sampling methods in Control API

bugmaker2 opened the pull request:
#### References to other Issues or PRs
An extension of @zolabar 's work on #24698 

#### Brief description of what is fixed or changed
1. Added a function named `gbt()` to discretize the transfer function. The functionality of pre-existed `bilinear()` and `backward_diff()` are integrated into the `gbt()`. 
2. Another method called `forward_diff()` was added. 

I believe merging them into a single function can improve the readability of the code. Also, more discretization methods can be added to this module.

**For users:**
`gbt()` and `forward_diff()` were added

**For developers:**
`bilinear()`, `forward_diff()` and `backward_diff()` are defined internally via `gbt()`.

Calling `bilinear(tf, Ts)` is the same as calling `gbt(tf, Ts, 0.5)`
`backward_diff(tf, Ts)` the same as `gbt(tf, Ts, 1)`
`forward_diff(tf, Ts)` the same as`gbt(tf, Ts, 0)`

Further, now, all methods return coefficients in normalized form (leading coefficient of denominator is 1).
$$H(z)=\sum_{i=0}^{n}a_i z^i  / \sum_{j=0}^{m} b_j z^j $$

$$H(z)=\sum_{i=0}^{n}\frac{a^i}{b_m}z^i / (1+\sum_{j=1}^{m} \frac{b_j}{b_m} z^j)$$

#### Other comments
More work needs to be done on the expansion of gbt(). We also expect there to be a `sample()` function developed within `TransferFuntion` which returns a discrete transfer function.

The reference for the implementation of `gbt()`
https://www.polyu.edu.hk/ama/profile/gfzhang/Research/ZCC09_IJC.pdf



We can take this doc as a reference as well
https://python-control.readthedocs.io/en/0.9.3.post2/generated/control.TransferFunction.html#control.TransferFunction.sample
The method includes `gbt()`, `bilinear()`, `euler()`, `backward_diff()` methods in sample().

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* physics.control
  * Add gbt and forware_diff in Control API
<!-- END RELEASE NOTES -->

Commit:
control: Added sample() and merged different methods

Added a function named sample to discretized transfer function. The pre-existed bilinear and backward_diff method were merged into this function. Another method called forward_diff was added into this function. I believe merging them into a single function can improve the readability of the code. Also more discretization methods can be added into this new function.

zolabar commented:
@bugmaker2 , thank you for the contribution.

what do you thing concerning backward compatibility? I think that we could keep the current usage and still include your Innovation.

for example, keep

bilinear(tf, T) and backward_diff(tf, T)

Add

 gbt(tf, T, alpha) 

and

forward_diff(tf, T)

and internally use gbt to define bilinear, backward_diff and forward_diff()

as

forward_diff(tf, T) = gbt(tf, T, 0)

bilinear(tf, T) = gbt(tf, T, 0.5)

backward_diff(tf, T) = gbt(tf, T, 1)

@hanspi42 , what do you think? Thank you!

zolabar commented:
@bugmaker2 , I just had a look at the code, please have a look at gbt and its relation to forward_diff, bilinear and backward_diff (some call bilinear also tustin)

Formula (3) in

https://www.polyu.edu.hk/ama/profile/gfzhang/Research/ZCC09_IJC.pdf

then in the coding, only implement the Add loop for gbt, and then just reuse gbt to define the others as I mentioned above

then no if and else are necessary and the similar code lines are not repeated 🙂

bugmaker2 commented:
As I am searching the dependencies on `bilinear()` and `backward_diff()`, I found there isn't any connection to other parts of SymPy, including the notebooks. So we may not need to consider backward compatibility but forward compatibility.

I understand that the difference between bilinear, forward & backward diff is just a matter of the value of alpha in gbt. I am just considering combining the different methods into one collection. So the code above is more like a showcase of my idea instead of a real implementation. I would try to minimize the repeat in the code as the idea is confirmed.
But in other words, it is a common practice to introduce discretization methods with the format of a single function:
[c2d in Matlab](https://www.mathworks.com/help/control/ref/dynamicsystem.c2d.html#mw_53fc4689-2099-41d0-93b3-de1e51a174c1)
[TransferFunction.sample in  Python Control Systems Library](https://python-control.readthedocs.io/en/0.9.3.post2/generated/control.TransferFunction.html)
So, can we introduce the sample() into the class TransferFunction as a static method?

What's more, I think the return value of `sample()` or `bilinear()` should be in the format of TransferFunction instead of lists of nums and dens. What do you think?

zolabar commented:
> As I am searching the dependencies on `bilinear()` and `backward_diff()`, I found there isn't any connection to other parts of SymPy, including the notebooks. So we may not need to consider backward compatibility but forward compatibility.
> 
> I understand that the difference between bilinear, forward & backward diff is just a matter of the value of alpha in gbt. I am just considering combining the different methods into one collection. So the code above is more like a showcase of my idea instead of a real implementation. I would try to minimize the repeat in the code as the idea is confirmed. But in other words, it is a common practice to introduce discretization methods with the format of a single function: [c2d in Matlab](https://www.mathworks.com/help/control/ref/dynamicsystem.c2d.html#mw_53fc4689-2099-41d0-93b3-de1e51a174c1) [TransferFunction.sample in Python Control Systems Library](https://python-control.readthedocs.io/en/0.9.3.post2/generated/control.TransferFunction.html) So, can we introduce the sample() into the class TransferFunction as a static method?
> 
> What's more, I think the return value of `sample()` or `bilinear()` should be in the format of TransferFunction instead of lists of nums and dens. What do you think?

Thank you for the feedback.

I think that we should separate some topics.

1.) The backward compatibility

in sympy 1.12 the functions bilinear() and backward_diff() return the coefficients of the discrete transfer functions. Hence, if users use this and they want their scripts to work in sympy 1.12.1 or 1.12.2 or 1.13, then we need to guarantee via backward compatibilty, that bilinear() and backward_diff() still exist and return the same result. However, internally we can implement them as particular cases of gbt(), which would mean only ONE code of the generalized Add(...) containing alpha.

2.) Matlab and scipy.signal use these kind of structures, too, i.e. discretization functions that return the coefficients, which is practical when dealing with real life control circuits

https://de.mathworks.com/help/signal/ref/bilinear.html

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.bilinear.html

3.) I think that it is a very good idea to have additionally the function sample(tf, T, method='...') that returns an object of a class TransferfunctionZ or discreteTransferFunction.

And here comes what I meant in the e-mail in the SymPy group :). That class that not exist yet! So you could start by creating it, by translating all the existing transfer function analysis in the Laplace variable to the discrete transfer function analysis in the Z-Transform variable "z".

Once that object exists, "sample" can get a s-TransferFuction and return an z-TransferFunction, which would have the corresponding properties in the z-TransferFunction world, pole plot etc... (stability if poles are in unit circle etc...)

moorepants commented:
> in sympy 1.12 the functions bilinear() and backward_diff() take s-TransferFunctions return the coefficients of the discrete transfer functions. Hence, if users use this and they want their scripts to work in sympy 1.12.1 or 1.12.2 or 1.13, then we need to guarantee via backward compatibilty, that bilinear() and backward_diff() still exist and return the same result. However, internally we can implement them as particular cases of gbt(), which would mean only ONE code of the generalized Add(...) containing alpha.

This is the right interpretation. If we make a release that includes public functions, then we have to honor backwards compatibility and follow a deprecation procedure if you must change it. In this case, there seems to be an easy solution to avoid changing the existing API so we shouldn't deprecate.

bugmaker2 commented:
Thanks for your explanation @zolabar @moorepants . I got what you mean by saying backward compatibility now. I may check [Deprecation Policy - Sympy](https://docs.sympy.org/latest/contributing/deprecations.html) before I do such a change.

Also, I believe the idea of creating a discrete TransferFunction class can be included in my GSoC proposal this year. But, as I can mimic TransferFunction now after @namannimmo10 's great work, I think the workload of just implementing this part is not enough for 350hrs. 

I noticed that @zolabar mentioned the cooperation between discrete and physics/control in #24985 . I am going to read the code of discrete.

bugmaker2 commented:
I have finished my implementation of `gbt()`, and I restored the `bilinear()`, `backward_diff()` methods with `forward_diff()` method.
The core code of `gbt()` looks like this:
```Python
num = Add(*[ T**(N-i) * c * (z-1)**i * (alpha * z + 1 - alpha)**(N-i) for c, i in zip(np[::-1], range(len(np))) ])
den = Add(*[ T**(N-i) * c * (z-1)**i * (alpha * z + 1 - alpha)**(N-i) for c, i in zip(dp[::-1], range(len(dp))) ])
```
And the code of `bilinear()` becomes this:
```Python
def bilinear(tf, sample_per):
    return gbt(tf, sample_per, 0.5)
```
And I tested the `bilinear()` method with the example in the docstring:
 ```cmd
>>> from sympy.physics.control.lti import TransferFunction, bilinear
>>> from sympy.abc import s, L, R, T
>>> tf = TransferFunction(1, s*L + R, s)
>>> numZ, denZ = gbt(tf, T, 0.5)
```
Instead of returning the correct answer,
```cmd
>>> numZ
[T, T]
>>> denZ
[2*L + R*T, -2*L + R*T]
```
The method returned me two lists with alpha multiplied, which triggered the error message in test:
```cmd
>>> numZ
[0.5*T, 0.5*T]
>>> denZ
[1.0*L + 0.5*R*T, -1.0*L + 0.5*R*T]
```

I tried to use `simplify()` to solve this problem:
```Python
from sympy import symbols, simplify
T = symbols('T')
lst = [1.0*T, 1.0*T]
simplified_lst = [simplify(expr).as_coefficients_dict()[T] for expr in last]
```
This returned me:
``` cmd
>>> simplified_lst
[1.00000000000000, 1.00000000000000]
```
And I also tried `Rational()`
```Python
from sympy import symbols, Rational
T = symbols('T')
lst = [1.0*T, 1.0*T]
rational_lst = [Rational(str(expr)) for expr in lst]
simplified_lst = [coeff.as_integer_ratio()[0] for coeff in rational_lst]
```
This returned me:
```cmd
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <listcomp>
  File "/Users/colourfulwhite/Codes/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/colourfulwhite/Codes/sympy/sympy/core/numbers.py", line 1623, in __new__
    raise TypeError('invalid input: %s' % p)
TypeError: invalid input: 1.0*T
```
Is there any way to fix this problem?

Commit:
Restore the bilinear and backward diff, create gbt and forward_diff

Commit:
Fix divided by zero bug. Main bug remains.

Commit:
Eliminate the error in flask8

Commit:
Fix doctest problem

Commit:
A rather ugly way to solve the normalization problem

bugmaker2 commented:
Hi, @zolabar . This is a rather ugly way to solve the normalization problem. I create a new fraction using the coefficients, then I use the built-in `nsimplify()` and `simplify()` functions to normalize the fraction (A single function does not fully normalize fractions with float coefficients).
```Python
# Simplify the fraction
    new_tf = nsimplify(new_num / new_den).simplify()
```
That is my solution for now. This solution returns the correct result with normalized coefficients.

zolabar commented:
> Hi, @zolabar . This is a rather ugly way to solve the normalization problem. I create a new fraction using the coefficients, then I use the built-in `nsimplify()` and `simplify()` functions to normalize the fraction (A single function does not fully normalize fractions with float coefficients).
> 
> ```python
> # Simplify the fraction
>     new_tf = nsimplify(new_num / new_den).simplify()
> ```
> 
> That is my solution for now. This solution returns the correct result with normalized coefficients.

Ok.

The aim is to do all this without any simplify function.

wouldn‘t it have been enough to divide all coefficients by the coefficient of the leading term of the denominator?

And to keep your  nice extension of the Add(…) part (extension by alpha)?

@hanspi42 had already developed the nice Add(…) part, to get rid of simpliy functions

bugmaker2 commented:
Here is the results from numerical counterparts:

**Matlab**:
```matlab
numerator = [1]; % The coefficients for the numerator of the transfer function
denominator = [1,2,1]; % The coefficients for the denominator of the transfer function

% Define the sampling frequency
sampling_frequency = 1000; % 1000 Hz

% Convert the continuous-time filter to a discrete-time filter using the bilinear transformation
[numerator_discrete, denominator_discrete] = bilinear(numerator, denominator, sampling_frequency);

Discrete-time filter numerator coefficients:
   0.9960   1.9920   0.9960
Discrete-time filter denominator coefficients:
   1.0000   1.9920   0.9920
```
The result from **Scipy** is slightly different with higher precision:
```Python
import numpy as np
from scipy.signal import bilinear

# Define the continuous-time filter coefficients
numerator = np.array([1])  # The coefficients for the numerator of the transfer function
denominator = np.array([1,2,1])  # The coefficients for the denominator of the transfer function

# Define the sampling frequency
sampling_frequency = 1/1000  # 1000 Hz

# Convert the continuous-time filter to a discrete-time filter using the bilinear transformation
numerator_discrete, denominator_discrete = bilinear(numerator, denominator, fs=sampling_frequency)

print("Discrete-time filter numerator coefficients:", numerator_discrete)
print("Discrete-time filter denominator coefficients:", denominator_discrete)

# Discrete-time filter numerator coefficients: [0.99601197 1.99202394 0.99601197]
# Discrete-time filter denominator coefficients: [1.         1.99201597 0.9920319 ]
```
What they actually do is, as you said, divide all coefficients by the coefficient of the leading term of the denominator.
$$H(z)=\sum_{i=0}^{n}a_i z^i  / \sum_{j=0}^{m} b_j z^j $$

$$H(z)=\sum_{i=0}^{n}\frac{a^i}{b_m}z^i / (1+\sum_{j=1}^{m} \frac{b_j}{b_m} z^j)$$

My concern is that the new format of results will raise backward compatibility issues as you mentioned before. The implementation is quite easy though:
```python
    para = den_coefs[0]
    num_coefs = [coef/para for coef in num_coefs]
    den_coefs = [coef/para for coef in den_coefs]
```
```ipython
>>> from sympy.physics.control.lti import TransferFunction, gbt
>>> from sympy.abc import s, L, R, T
>>> tf = TransferFunction(1, s*L + R, s)
>>> numZ, denZ = gbt(tf, T, 0.5)
>>> numZ
[0.5*T/(1.0*L + 0.5*R*T), 0.5*T/(1.0*L + 0.5*R*T)]
>>> denZ
[1, (-1.0*L + 0.5*R*T)/(1.0*L + 0.5*R*T)]
```
Here are two main issues:
1. The new format of result is different from the original format of the result, which would raise backward compatibility issues:
```python
>>> numZ
[T, T]
>>> denZ
[2*L + R*T, -2*L + R*T]
```
2. Divide all coefficients by the coefficient of the leading term of the denominator would lead to Floating-point coefficient that cannot be removed, the `1.0` in the result, for example.

zolabar commented:
> Thanks for your explanation. Here is the results from numerical counterparts:
> 
> **Matlab**:
> 
> ```matlab
> numerator = [1]; % The coefficients for the numerator of the transfer function
> denominator = [1,2,1]; % The coefficients for the denominator of the transfer function
> 
> % Define the sampling frequency
> sampling_frequency = 1000; % 1000 Hz
> 
> % Convert the continuous-time filter to a discrete-time filter using the bilinear transformation
> [numerator_discrete, denominator_discrete] = bilinear(numerator, denominator, sampling_frequency);
> 
> Discrete-time filter numerator coefficients:
>    0.9960   1.9920   0.9960
> Discrete-time filter denominator coefficients:
>    1.0000   1.9920   0.9920
> ```
> 
> The result from **Scipy** is slightly different with higher precision:
> 
> ```python
> import numpy as np
> from scipy.signal import bilinear
> 
> # Define the continuous-time filter coefficients
> numerator = np.array([1])  # The coefficients for the numerator of the transfer function
> denominator = np.array([1,2,1])  # The coefficients for the denominator of the transfer function
> 
> # Define the sampling frequency
> sampling_frequency = 1/1000  # 1000 Hz
> 
> # Convert the continuous-time filter to a discrete-time filter using the bilinear transformation
> numerator_discrete, denominator_discrete = bilinear(numerator, denominator, fs=sampling_frequency)
> 
> print("Discrete-time filter numerator coefficients:", numerator_discrete)
> print("Discrete-time filter denominator coefficients:", denominator_discrete)
> 
> # Discrete-time filter numerator coefficients: [0.99601197 1.99202394 0.99601197]
> # Discrete-time filter denominator coefficients: [1.         1.99201597 0.9920319 ]
> ```
> 
> What they actually do is, as you said, divide all coefficients by the coefficient of the leading term of the denominator. H(z)=∑i=0naizi/∑j=0mbjzj
> 
> H(z)=∑i=0naibmzi/(1+∑j=1mbjbmzj)
> 
> My concern is that the new format of results will raise backward compatibility issues as you mentioned before. The implementation is quite easy though:
> 
> ```python
>     para = den_coefs[0]
>     num_coefs = [coef/para for coef in num_coefs]
>     den_coefs = [coef/para for coef in den_coefs]
> ```
> 
>  ```
> >>> from sympy.physics.control.lti import TransferFunction, gbt
> >>> from sympy.abc import s, L, R, T
> >>> tf = TransferFunction(1, s*L + R, s)
> >>> numZ, denZ = gbt(tf, T, 0.5)
> >>> numZ
> [0.5*T/(1.0*L + 0.5*R*T), 0.5*T/(1.0*L + 0.5*R*T)]
> >>> denZ
> [1, (-1.0*L + 0.5*R*T)/(1.0*L + 0.5*R*T)]
> ```
> 
> Here are two main issues:
> 
> 1. The new format of result is quite different from the original format of the result, which would raise backward compatibility issues:
> 
> ```python
> >>> numZ
> [T, T]
> >>> denZ
> [2*L + R*T, -2*L + R*T]
> ```
> 
>  2. Divide all coefficients by the coefficient of the leading term of the denominator without `simplify()` would lead to Floating-point coefficients that cannot be removed, the `1.0` in the result, for example.
>    What do you think? @zolabar

@bugmaker2 , thank you for the detailed analysis! And the numerical examples and the simple example of normalization using the current structure.

From my point of view a new result as

> >>> numZ
> [0.5*T/(1.0*L + 0.5*R*T), 0.5*T/(1.0*L + 0.5*R*T)]
> >>> denZ
> [1, (-1.0*L + 0.5*R*T)/(1.0*L + 0.5*R*T)]

is backward compatible enough. 
That means that every use case I can imagine would still work.

The new returned coefficients should yield the same results as before if used in scipy.signal.lfilter or similar.

May be the floating points vanish if you call gbt as

gbt(tf, T, 1/2)

?

For me, this is the preferred way.

Simple code, no use of simplify, and analog result structure as the numerical counterparts.

One would only still have to adapt and extend the test cases in lti_test.py

@hanspi42 , what is your Assessment of the current discretization extensions?

* include gbt (generalized bilinear transform)
* bilinear as gbt with alpha 1/2
* backward_diff as gbt with alpha 1
* include forward_diff as gbt with alpha 0
* Normalization of discrte tf, i.e. leading coeff of den is 1, like in literature and in the numerical counterparts

bugmaker2 commented:
The issue remains when I use `1/2`
> May be the floating points vanish if you call gbt as
gbt(tf, T, 1/2)
?
---

Another solution is to add `alpha = Rational(alpha)` in the code. It returns a perfect result when `alpha=0.5`
```python
>>> numZ, denZ = gbt(tf, T, 1/2)
>>> numZ
[T/(2*(L + R*T/2)), T/(2*(L + R*T/2))]
>>> denZ
[1, (-L + R*T/2)/(L + R*T/2)]
```
However, it returns an awful result when `alpha` is set to an unusual value:
```python
>>> numZ, denZ = gbt(tf, T, 0.2)
>>> numZ
[5404319552844595*T/(18014398509481984*(L + 5404319552844595*R*T/18014398509481984)), 12610078956637389*T/(18014398509481984*(L + 5404319552844595*R*T/18014398509481984))]
>>> denZ
[1, (-L + 12610078956637389*R*T/18014398509481984)/(L + 5404319552844595*R*T/18014398509481984)]
```
Then I found that `Rational(0.3) = 5404319552844595/18014398509481984` in Sympy. There is a precision problem in `Rational()`. The problem can be solved using `Rational(0.3).limit_denominator(1000)`. But I am afraid setting fixed limit would cause other problems. What do you think? @zolabar 

By the way, the `gbt()` function does well **without** `alpha=Rational(alpha)` when `alpha=0.3`, if you don't mind the float-point coefficients:
```python
>>> numZ, denZ = gbt(tf, T, 0.3)
>>> numZ
[0.3*T/(1.0*L + 0.3*R*T), 0.7*T/(1.0*L + 0.3*R*T)]
>>> denZ
[1, (-1.0*L + 0.7*R*T)/(1.0*L + 0.3*R*T)]
```

zolabar commented:
> The issue remains when I use `1/2`
> 
> > May be the floating points vanish if you call gbt as
> > gbt(tf, T, 1/2)
> > ?
> 
> Another solution is to add `alpha = Rational(alpha)` in the code. It returns a perfect result when `alpha=0.5`
> 
> ```python
> >>> numZ, denZ = gbt(tf, T, 1/2)
> >>> numZ
> [T/(2*(L + R*T/2)), T/(2*(L + R*T/2))]
> >>> denZ
> [1, (-L + R*T/2)/(L + R*T/2)]
> ```
> 
> However, it returns an awful result when `alpha` is set to an unusual value:
> 
> ```python
> >>> numZ, denZ = gbt(tf, T, 0.2)
> >>> numZ
> [5404319552844595*T/(18014398509481984*(L + 5404319552844595*R*T/18014398509481984)), 12610078956637389*T/(18014398509481984*(L + 5404319552844595*R*T/18014398509481984))]
> >>> denZ
> [1, (-L + 12610078956637389*R*T/18014398509481984)/(L + 5404319552844595*R*T/18014398509481984)]
> ```
> 
> Then I found that `Rational(0.3) = 5404319552844595/18014398509481984` in Sympy. There is a precision problem in `Rational()`. The problem can be solved using `Rational(0.3).limit_denominator(1000)`. But I am afraid setting fixed limit would cause other problems. What do you think? @zolabar
> 
> By the way, the `gbt()` function does well **without** `alpha=Rational(alpha)` when `alpha=0.3`, if you don't mind the float-point coefficients:
> 
> ```python
> >>> numZ, denZ = gbt(tf, T, 0.3)
> >>> numZ
> [0.3*T/(1.0*L + 0.3*R*T), 0.7*T/(1.0*L + 0.3*R*T)]
> >>> denZ
> [1, (-1.0*L + 0.7*R*T)/(1.0*L + 0.3*R*T)]
> ```

I don‘t mind the floating points, I think that one can keep it simple.

Commit:
vanish floating point coefficients

bugmaker2 commented:
Think I am ready for the merge. One last thing is the choice between float-point representation and fraction representation.
The difference is wheather add an `alpha = Rational(alpha).limit_denominator(1000)` in `gbt()`.

1. Add the `alpha = Rational(alpha).limit_denominator(1000)`
```python
>>> numZ, denZ = gbt(tf, T, 0.3)
>>> numZ
[3*T/(10*(L + 3*R*T/10)), 7*T/(10*(L + 3*R*T/10))]
>>> denZ
[1, (-L + 7*R*T/10)/(L + 3*R*T/10)]
```
2. Without `Rational(alpha)`
```python
>>> numZ, denZ = gbt(tf, T, 0.3)
>>> numZ
[0.3*T/(1.0*L + 0.3*R*T), 0.7*T/(1.0*L + 0.3*R*T)]
>>> denZ
[1, (-1.0*L + 0.7*R*T)/(1.0*L + 0.3*R*T)]
```

zolabar commented:
I think, I would skip the Rational function.

zolabar commented:
> Think I am ready for the merge. One last thing is the choice between float-point representation and fraction representation. The difference is wheather add an `alpha = Rational(alpha).limit_denominator(1000)` in `gbt()`.
> 
> 1. Add the `alpha = Rational(alpha).limit_denominator(1000)`
> 
> ```python
> >>> numZ, denZ = gbt(tf, T, 0.3)
> >>> numZ
> [3*T/(10*(L + 3*R*T/10)), 7*T/(10*(L + 3*R*T/10))]
> >>> denZ
> [1, (-L + 7*R*T/10)/(L + 3*R*T/10)]
> ```
> 
>  2. Without `Rational(alpha)`
> 
> ```python
> >>> numZ, denZ = gbt(tf, T, 0.3)
> >>> numZ
> [0.3*T/(1.0*L + 0.3*R*T), 0.7*T/(1.0*L + 0.3*R*T)]
> >>> denZ
> [1, (-1.0*L + 0.7*R*T)/(1.0*L + 0.3*R*T)]
> ```

@bugmaker2 , Please check the documentation of bilinear and and backward_diff , the example part od shifted

https://output.circle-artifacts.com/output/job/18050d76-e4d3-444b-b724-cd18f20aee1c/artifacts/0/doc/_build/html/modules/physics/control/lti.html#sympy.physics.control.lti.bilinear

In the „all checks have passed“ there is a documentation preview. 
Also, add gbt there, and forward_diff.

@moorepants, from my point of view (mathematical/Control theory perspective, this contribution is ok. After the question on the rational alpha and the documentation it should be ready.
Who is the reviewer of this PR? Thank you.

Commit:
without Rational()

zolabar commented:
@bugmaker2 , the „ Brief description of what is fixed or changed“ at the beginning of this PR are still dealing with the sample function (which can be added in an other PR, to return tfz once tfz exists as object)

Commit:
try to fix document problem

Commit:
trying to fix document

Commit:
hope this can do

zolabar commented:
@bugmaker2 ,

concerning documentation docstrings for new functions (gbt and forward_diff)

You need to add a directive ..autofunction to doc/src/modules/physics/control/lti.rst

Commit:
document fix

zolabar commented:
> #### References to other Issues or PRs
> An extension of @zolabar 's work on #24698
> 
> #### Brief description of what is fixed or changed
> Added a function named `gbt()` to discretize the transfer function. The pre-existed `bilinear()` and `backward_diff()` methods were merged into this function. Another method called `forward_diff()` was added to `gbt()`. I believe merging them into a single function can improve the readability of the code. Also, more discretization methods can be added to this new function. Before: `bilinear(tf, Ts)` `backward_diff(tf, Ts)` After: `gbt(tf, Ts, 0.5)` `gbt(tf, Ts, 0)` `gbt(tf, Ts, 1)`
> 
> #### Other comments
> More work needs to be done on the expansion of gbt(). We can take this doc as a reference https://python-control.readthedocs.io/en/0.9.3.post2/generated/control.TransferFunction.html#control.TransferFunction.sample The method includes gbt, bilinear, euler, backward_diff methods in sample().
> 
> #### Release Notes
> * physics.control
>   
>   * Merged bilinear and backward_diff in Control API

As I understand, bilinear and backward_diff still exist and can be used as before.

for users:

gbt and forward_diff were added

for developers:

internally bilinear, forward_diff and backward_diff are defeined via gbt

further, now, all methods return coefficients in normalized form (leading coefficient of denominator is 1).

zolabar commented:
@bugmaker2 , 

Good idea to latex the formulae of H(z), have a look at the docs if it is rendered as

$$H(z)=\frac{az+b}{cz+d}$$

Commit:
direct new funtions to sphinx

zolabar commented:
@moorepants , @sylee957 , @hanspi42 , from my point of view (mathematical/Control theory perspective) this contribution is ok.
Who can review and merge this PR? Thank you.

sylee957 commented on sympy/physics/control/lti.py:
```diff
@@ -33,90 +32,171 @@ def _roots(poly, var):
+    >>> denZ
+    [1, (-1.0*L + 0.5*R*T)/(1.0*L + 0.5*R*T)]
+    """
+    return gbt(tf, sample_per, 0.5)
```
It is better to use the symbolic rationals than floats
```suggestion
    return gbt(tf, sample_per, S.Half)
```

sylee957 commented on sympy/physics/control/lti.py:
```diff
@@ -33,90 +32,171 @@ def _roots(poly, var):
+    >>> denZ
+    [1, (-L + R*T)/L]
+    """
+    return gbt(tf, sample_per, 0)
```
```suggestion
    return gbt(tf, sample_per, S.Zero)
```

sylee957 commented on sympy/physics/control/lti.py:
```diff
@@ -33,90 +32,171 @@ def _roots(poly, var):
+    >>> denZ
+    [1, -L/(L + R*T)]
+    """
+    return gbt(tf, sample_per, 1)
```
```suggestion
    return gbt(tf, sample_per, S.One)
```

bugmaker2 committed:
Apply suggestions from code review

The S really solves the representation issue. But pass `alpha=0.3` to `gbt()` will cause floating points too. So, maybe we should consider using `Rational()`

Co-authored-by: S.Y. Lee  <sylee957@gmail.com>

Commit:
finally use fraction instead of float point representation

Commit:
update mailmap

sylee957 commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -1222,24 +1222,73 @@ def test_TransferFunctionMatrix_functions():
+    tf_test_manual = TransferFunction(T/a, s + (-a + b*T)/a, s)
 
-    assert S.Zero == (tf_test_bilinear-tf_test_manual).simplify().num
+    assert S.Zero == (tf_test_forward.simplify()-tf_test_manual.simplify()).simplify().num
```
Why doesn't `(tf_test_forward-tf_test_manual).simplify()` work?

sylee957 commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -1222,24 +1222,73 @@ def test_TransferFunctionMatrix_functions():
+    assert S.Zero == (tf_test_backward.simplify()-tf_test_manual.simplify()).simplify().num
+
+    tf = TransferFunction(1, a*s+b, s)
+    numZ, denZ = gbt(tf, T, 0.3)
```
It is also better to use exact rationals here

```suggestion
    numZ, denZ = gbt(tf, T, Rational(3, 10))
```

bugmaker2 replied:
This is designed to simulate random user input. Users do not necessarily use `Rational()` functions as input, but rather give more arbitrary floating-point values. Since I wrote the `Rational(alpha)` transformation into the `gbt()` function, this probably won't be a problem, I think.

zolabar commented:
@bugmaker2 thank you for the contribution!

As a next point and larger topic is the creation of a TransferFunctionZ class (or similar name) with corresponding methods and properties of z-transfer function analysis, i.e. discrete signal processing/control. @hanspi42 , what do you think?

@sylee957 , thank you for the review and merge!

hanspi42 commented:
> As a next point and larger topic I see the creation of a TransferFunctionZ class (or similar name) with corresponding methods and properties of z-transfer function analysis, i.e. discrete signal processing/control. @hanspi42 , what do you think?

Good idea. Please note that @edelrosespol presently works towards implementing the z transform itself in #24985. Maybe there are synergies there?

zolabar commented:
> 

@hanspi42 , thank you for the asessment.
I also see synergies there. 
However, I think that this is a larger topic. It could need some previous road map discussion :)

bugmaker2 commented:
@zolabar @sylee957 @moorepants @hanspi42 Thank you all for your suggestions and help. 

> As a next point and larger topic I see the creation of a TransferFunctionZ class (or similar name) with corresponding methods and properties of z-transfer function analysis

Because the `gbt()` function is relatively independent, changes are not difficult. I changed this part mainly to familiarize myself with the contribution process. If I want to work on a new class, I need to look at the code as a whole for some time.

> Please note that @edelrosespol presently works towards implementing the z transform itself in https://github.com/sympy/sympy/pull/24985. Maybe there are synergies there?

Yes, I noticed this PR #24985 . Thank you for the reminder.

hanspi42 commented:
> @hanspi42 , thank you for the asessment. I also see synergies there. However, I think that this is a larger topic. It could need some previous road map discussion :)

Let's discuss it here: #25040

oscarbenjamin commented on .mailmap:
```diff
@@ -365,6 +365,8 @@ Ayush Malik <ayushmalik779@gmail.com> Ayush-Malik <ayushmalik779@gmail.com>
 Ayushman Koul <ayushmankoul4570@gmail.com>
 B McG <bmcg0890@gmail.com>
+Baiyuan Qiu <1061688677@qq.com> <qby1061688677@gmail.com>
+Baiyuan Qiu <51898470+bugmaker2@users.noreply.github.com>
```
You should not add multiple lines to the .mailmap file like this:
https://github.com/sympy/sympy/blob/a775270e14bdedc4a39b4fb9c6e4e6e398be896f/.mailmap#L125-L148
@sylee957 can you check for this when merging PRs?
